pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,1438 @@
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/golang/glog"
+)
+
+// Design:
+//
+// The fundamental key to this design is the bi-directional "pointer" between
+// PersistentVolumes (PVs) and PersistentVolumeClaims (PVCs), which is
+// represented here as pvc.Spec.VolumeName and pv.Spec.ClaimRef. The bi-directionality
+// is complicated to manage in a transactionless system, but without it we
+// can't ensure sane behavior in the face of different forms of trouble.  For
+// example, a rogue HA controller instance could end up racing and making
+// multiple bindings that are indistinguishable, resulting in potential data
+// loss.
+//
+// This controller supports pre-bound (by the creator) objects in both
+// directions: a PVC that wants a specific PV or a PV that is reserved for a
+// specific PVC.
+//
+// The binding is two-step process. PV.Spec.ClaimRef is modified first and
+// PVC.Spec.VolumeName second. At any point of this transaction, the PV or PVC
+// can be modified by user or other controller or completelly deleted. Also, two
+// (or more) controllers may try to bind different volumes to different claims
+// at the same time. The controller must recover from any conflicts that may
+// arise from these conditions.
+
+// annWasEverBound annotation applies to PVCs.  It indicates that the lifecycle
+// of the PVC has passed through the initial setup.  This information changes
+// how we interpret some observations of the state of the objects.
+const annWasEverBound = "pv.kubernetes.io/bound-completed"
+
+// annBoundByController annotation applies to PVs and PVCs.  It indicates that
+// the binding (PV->PVC or PVC->PV) was installed by the controller.  The
+// absence of this annotation means the binding was done by the user (i.e.
+// pre-bound).
+const annBoundByController = "pv.kubernetes.io/bound-by-controller"
+
+// annClass annotation represents a new field which instructs dynamic
+// provisioning to choose a particular storage class (aka profile).
+const annClass = "volume.alpha.kubernetes.io/storage-class"
+
+// This annotation is added to a PV that has been dynamically provisioned by
+// Kubernetes. It's value is name of volume plugin that created the volume.
+// It serves both user (to show where a PV comes from) and Kubernetes (to
+// recognize dynamically provisioned PVs in its decissions).
+const annDynamicallyProvisioned = "pv.kubernetes.io/provisioned-by"
+
+// Name of a tag attached to a real volume in cloud (e.g. AWS EBS or GCE PD)
+// with namespace of a persistent volume claim used to create this volume.
+const cloudVolumeCreatedForClaimNamespaceTag = "kubernetes.io/created-for/pvc/namespace"
+
+// Name of a tag attached to a real volume in cloud (e.g. AWS EBS or GCE PD)
+// with name of a persistent volume claim used to create this volume.
+const cloudVolumeCreatedForClaimNameTag = "kubernetes.io/created-for/pvc/name"
+
+// Name of a tag attached to a real volume in cloud (e.g. AWS EBS or GCE PD)
```

Comment:
I guess clouds can internally map this to other names - these strings really define an API between the controller and the volume plugins.  In effect they aren't arbitrary tags.  Maybe we should make them fields in options, and let volume plugins turn them into tag names?